### PR TITLE
Fix executables linking

### LINKED ISSUE #556: Detect submodule dependencies
If a procedure inside an `example/` subdirectory has an interface in a `module` and the corresponding procedure definition in a `submodule` (which was largely the intent when `submodule` became a language feature in Fortran 2008), then a linking error results.  It appears that `fpm` is only seeing the `module` dependency information captured in `use` statements but not in `submodule` statements.

To Do: 
- [ ] post a reproducer later.

### PR DESCRIPTION
**Pull request by LKedward**
- Adds: new test packages for apps with non-module non-library source files (`app_with_c` and `app_with_submodule`)
- Adds: new routine to collect non-module non-library source targets as link dependencies of the respective exectutables
  - Only link those sources in the same directory and in a subdirectory of the executable source file

Fixes #556 and #741

### CODE REVIEW

File: src/fpm_targets.f90 [hunk 1]
```diff
@@ -334,6 +336,57 @@ end function get_object_name
...
+                associate(exe => targets(j)%ptr, dep => targets(i)%ptr)
+
+                    exe_source_dir = dirname(exe%dependencies(1)%ptr%source%file_name)
+
+                    if (allocated(dep%source)) then
+
+                        if (dep%source%unit_scope /= FPM_SCOPE_LIB .and. &
+                            dep%source%unit_type /= FPM_UNIT_PROGRAM .and. &
+                            dep%source%unit_type /= FPM_UNIT_MODULE .and. &
+                            index(dirname(dep%source%file_name), exe_source_dir) == 1) then
```
Comment (zoziha) [inline src/fpm_targets.f90:371]:
````fortran
subroutine a
end subroutine a
module b_m
end module b_m
````
It seems that we still can't handle the case of Fortran source files mixed with `non-module procedure` and `module`? However, this mixed writing is not a good coding style, and we can ignore it for now.

**Review by zoziha — APPROVED** [review]
Thanks for such a quick fix, @LKedward .
This patch passed all tests, LGTM.

File: src/fpm_targets.f90 (hunk 1 above)
Comment (LKedward) [reply src/fpm_targets.f90:371]: Actually that case should still work. If a source file contains a non-module procedure then it is marked as `FPM_UNIT_SUBPROGRAM` not `FPM_UNIT_MODULE` as per the rules needed for tree-shaking, see [here](https://fortran-lang.github.io/fpm/module/fpm_model.html).

**Review by arteevraina — APPROVED** [review]
Great work! This also fixes the `test_hash_function` linker issue we are facing in #729

**Review by awvwgk — APPROVED** [review]
Thanks for the quick fix. Looks good to me.

**Comment by LKedward** [comment]
Thanks all for reviewing, I'll now merge.

**Pull request merged**